src/hotspot/os/linux/malloctrace/malloc_trace.md

```diff
@@ -0,0 +1,140 @@
+(c) SAP 2021
+
+# The SapMachine MallocTrace facility
+
+## Preface
+
+When analyzing native OOMs resulting from C-heap exhaustion, we have two facilities built into the VM:
+
+- NMT (available in OpenJDK, SapMachine, and SAP JVM)
+- the SAP JVM malloc statistics (as the name indicates, SAP JVM only)
+
+Both facilities have pros and cons, but they share one significant disadvantage: they require the malloc sites to be instrumented at compile time. Therefore they are unsuited for analyzing situations where outside code within the VM process allocates C-heap.
+
+In the past, if we were facing a suspected C-heap leak from code outside the VM (outside the coverage of NMT or SAP JVM malloc statistics), we were helpless. In these situations, one would typically use tools like perf or Valgrind, but that is seldom an option for us.
+
+### mtrace?
+
+A straightforward possibility to do this would be the Glibc-internal trace. Using `mtrace(3)`, one can force the Glibc to write a trace file for all malloc calls. Unfortunately, it is very costly. In my experiments, VM slowed down by factor 8-10. It also has to be enabled at the start of the VM (when the VM still happens to be single-threaded). That usually rules it out in production scenarios.
+
+## The SapMachine MallocTrace facility
+
+The new Linux-only malloc trace facility in the SapMachine uses Glibc malloc hooks to hook into the allocation process. In that way, it resembles the Glibc-internal `mtrace(3)`. But unlike `mtrace(3)`, it accumulates data in memory and provides a condensed report upon request, making it a lot faster. Moreover, it does not have to be started at VM startup (though it can), but one can switch it on when needed.
+
+### Usage via jcmd
+
+#### Switch trace on:
+
+```
+thomas@starfish$ jcmd AllocCHeap System.malloctrace on
+268112:
+Tracing activated
+```
+
+#### Switch trace off 
+```
+thomas@starfish$ jcmd AllocCHeap System.malloctrace off
+268112:
+Tracing deactivated
+```
+
+#### Print a SapMachine MallocTrace report. 
```

Comment:
Same here:

#### Print a SapMachine MallocTrace report:

(use colon at the end, like above)